train/chess.py

```diff
@@ -0,0 +1,105 @@
+
+import json
+
+import numpy as np
+import tensorflow as tf
+from keras import optimizers, Input
+from keras.layers import Dense
+from keras.models import Model
+import keras
+
+from trainable_game import TrainableGame
+import net_utils
+
+_LEARNING_RATE = 0.001
+
+
+class NetType:
+    SimpleTwoHeaded = "simple_two_headed"
+    ConvNetV1 = "ConvNetV1"
+
+
+class Chess(TrainableGame):
+    BOARD_SIZE = 8
+    PLANES_NUM = 18
+    MOVE_NUM = 1880
+
+    def load_data_entry(self, path):
+        with open(path, "rb") as f:
+            data_entry = json.load(f)
+
+        planes = np.array(data_entry["planes"], dtype=np.uint64)
+        planes = np.frombuffer(planes, dtype=np.uint32).reshape((self.PLANES_NUM, 2))
+
+        probs = np.array(data_entry["probs"], dtype=np.float32)
+        assert len(probs) == self.MOVE_NUM
+
+        winner = np.float32(data_entry["winner"])
+
+        return (planes, probs, winner)
+
+    def _get_input_shape(self, cfg):
+        shape_cpu = (self.BOARD_SIZE, self.BOARD_SIZE, self.PLANES_NUM)
+        shape_gpu = (self.PLANES_NUM, self.BOARD_SIZE, self.BOARD_SIZE)
+        return shape_cpu if cfg["cpu"] else shape_gpu
+
+    def _create_model_simple_two_headed(self, cfg):
+        inputs = Input(
+            shape=self._get_input_shape(cfg),
+            name="input_planes")
+        flow = tf.keras.layers.Flatten()(inputs)
+        x = Dense(units=128, activation="relu")(flow)
+        head_val = Dense(
+            units=1, activation="tanh", name="value_head")(x)
+        head_probs = Dense(units=self.MOVE_NUM, name="policy_head")(x)
+        model = Model(inputs=inputs, outputs=[head_val, head_probs])
+
+        opt = optimizers.Adam(learning_rate=_LEARNING_RATE)
+        model.compile(
+            optimizer=opt,
+            loss={'value_head': tf.keras.losses.MeanSquaredError(),
+                  'policy_head': net_utils.loss_cross_entropy},
+            metrics={'value_head': tf.keras.metrics.MeanSquaredError(),
+                     'policy_head': net_utils.policy_head_accuracy})
+        return model
+
+    def _create_model_convnetv1(self, cfg):
+        inputs = Input(
+            shape=self._get_input_shape(cfg),
+            name="input_planes")
+        outputs = net_utils.create_convnetv1(
+            inputs,
+            residual_filter_num=cfg["model"]["residual_filter_num"],
+            residual_block_num=cfg["model"]["residual_block_num"],
+            moves_num=self.MOVE_NUM,
+            l2reg=cfg["model"]["l2reg"],
+            cpu=cfg["cpu"])
+        model = Model(inputs=inputs, outputs=outputs)
+
+        opt = optimizers.Adam(learning_rate=_LEARNING_RATE)
+        model.compile(
+            optimizer=opt,
+            loss={'value_head': tf.keras.losses.MeanSquaredError(),
+                  'policy_head': net_utils.loss_cross_entropy},
+            metrics={'value_head': tf.keras.metrics.RootMeanSquaredError(),
+                     'policy_head': net_utils.policy_head_accuracy})
```

Comment:
how do you choose the metrics? they are just for human-reading right? they don't affect the training process

Reply:
yes, but they are the best indication we have. I also want to open a ticket of collecting the return value of model.fit, which return some sort of history with the metrics. This will give us a lot of debug information